tempesta_fw/cfg.h

```diff
@@ -0,0 +1,285 @@
+struct TfwCfgSpec {
+	const char *name;
+	const char *deflt;
+	int (*handler)(TfwCfgSpec *self, TfwCfgEntry *parsed_entry);
+	void *dest;
+	void *spec_ext;		/* TfwCfgSpecInt, TfwCfgSpecStr, etc. */
+	int call_counter;	/* calls of @handler */
+	bool allow_repeat;	/* allow call_counter > 1 */
+	bool allow_none;	/* allow call_counter == 0 */
+};
+
+/**
+ * Walks over a NULL-terminated array of TfwCfgSpec structures.
+ */
+#define TFW_CFG_FOR_EACH_SPEC(spec_pos, spec_arr) \
+	for ((spec_pos) = (spec_arr); (spec_pos)->name; ++(spec_pos))
+
+/**
+ * Internally, the Tmpesta FW code is split into modules (not kernel modules,
+ * but rather loosely coupled pieces of code). We need to distribute parsed
+ * configuration data and start/stop events over them, but we don't want to make
+ * the parser depend on all possible modules, so we utilize the late binding
+ * approach here.
+ *
+ * Each module defines a TfwCfgMod structure and calls tfw_cfg_mod_register().
+ * We maintain a list of registered modules in runtime and push events and
+ * configuration data via callbacks.
+ *
+ * @name is a unique text identifier of the module.
+ * Just like C identifiers, it must consist of alphanumeric characters and start
+ * with a letter and so on.
+ *
+ * @specs is the specification for the configuration parser that lists all
+ * possible configuration sections and directives for this module and describes
+ * how to handle them.
+ * @specs must be an array of TfwCfgSpec structures which is terminated
+ * by a null (zero'ed) element.
+ *
+ *
+ * @init and @exit callbacks act like module_init()/module_exit().
+ * They are called when a module is registered/unregistered and their only
+ * purpose is to eliminate some boilerplate function declarations and calls.
+ *
+ * @setup, @start, @stop and @cleanup callbacks are invoked when events
+ * are recevied via sysctl. The workflow is the following:
+ *   1. Receive a "start" event via syctl.
+ *   2. Invoke @setup.
+ *   3. Parse configuration, process @specs.
+ *   4. Invoke @start.
+ *   5. Receive a "stop" event via sysctl.
+ *   6. Invoke @stop.
+ *   7. Invoke @cleanup.
+ * Separate @setup/@start are needed to be able to run actions both before
+ * and after configuration parsing.
+ * Separate @stop/@cleanup callbacks are needed to avoid a scenario when a
+ * half of the modules are stopped and free()'d their memory, and another half
+ * is working and possibly references memory of the stopped modules. We stop
+ * work in @stop and free memory in @cleanup to avoid tricky synchronization.
+ *
+ * The only required field is the @name. All other fields may be NULL.
+ */
+typedef struct {
+	struct list_head list;	/* Private. Don't touch. */
+	const char *name;	/* [A-Za-z0-9_], starts with a letter. */
+	TfwCfgSpec *specs;	/* An array terminated by a null element. */
+	int  (*init)(void);
+	void (*exit)(void);
+	int  (*setup)(void);
+	int  (*start)(void);
+	void (*stop)(void);
+	void (*cleanup)(void);
+} TfwCfgMod;
+
```

Comment:
Yeah... But I'm still confused by the fact that any of the existing modules doesn't use setup and start callbacks at the same time. So it seems that they can be aggregates. Also why do we need stop and cleanup at the same time? Why we shouldn't call cleanup routine at the stop?

Meantime, it's obvious that we need to separate init/exit from start/stop. However, why setup/cleanup are separate from start/stop is not clear. Please explain.